upup/pkg/fi/cloudup/dotasks/droplet.go

```diff
@@ -0,0 +1,136 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dotasks
+
+import (
+	"context"
+	"strconv"
+
+	"github.com/digitalocean/godo"
+
+	"k8s.io/kops/pkg/resources/digitalocean"
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kops/upup/pkg/fi/cloudup/do"
+	_ "k8s.io/kops/upup/pkg/fi/cloudup/terraform"
+)
+
+//go:generate fitask -type=Droplet
+type Droplet struct {
+	Name      *string
+	ID        *string
+	Lifecycle *fi.Lifecycle
+
+	Region   *string
+	Size     *string
+	Image    *string
+	SSHKey   *string
+	Tags     []string
+	UserData *fi.ResourceHolder
+}
+
+var _ fi.CompareWithID = &Droplet{}
+
+func (d *Droplet) CompareWithID() *string {
+	return d.ID
```

Comment:
CompareWithID is used to identify two objects that are the same thing, in particular when an object refers to another object.  e.g. If Image was actually an object here, we might have two Droplet objects that both set a stub Image, and we might define a task for that Image, and we want to match all 3.  CompareWithID is what is matched to match them.

The reason we care about reconciling the duplicates is primarily for dependency analysis, so we can know to run the Image task before the Droplet task in this example.

So in this case, you probably want to compare by Name, because that is what you would set in a stub Droplet I would guess - in particular you're unlikely to know the ID.  But in practice, it probably doesn't matter anyway, unless you're actually creating stub Droplets in another task.